wagtail/wagtailcore/models.py

```diff
@@ -40,8 +39,86 @@
 
 PAGE_TEMPLATE_VAR = 'page'
 
+SITE_CACHE = {}
+REQUEST_SITE_MATCH_CACHE = {}
+
 
 class SiteManager(models.Manager):
+
+    @staticmethod
+    def get_hostname_and_port_from_request(request):
+        """Reliably extract 'hostname' and 'port' values from the supplied
+        ``HttpRequest`` instance (without raising exceptions), and return them
+        as a tuple.
+        """
+        try:
+            hostname = request.get_host().split(':')[0]
+        except KeyError:
+            hostname = None
+        try:
+            port = request.get_port()
+        except KeyError:
+            port = None
+        return hostname, port
+
+    @staticmethod
+    def clear_site_cache():
+        global SITE_CACHE
+        global REQUEST_SITE_MATCH_CACHE
+        SITE_CACHE = {}
+        REQUEST_SITE_MATCH_CACHE = {}
+
+    def _get_by_id(self, site_id):
+        if site_id not in SITE_CACHE:
+            site = self.get(pk=site_id)
+            SITE_CACHE[site_id] = site
+        return SITE_CACHE[site_id]
+
+    def _cache_and_return_match(self, site, match_cache_key):
+        SITE_CACHE[site.pk] = site
+        REQUEST_SITE_MATCH_CACHE[match_cache_key] = site.pk
```

Comment:
I don’t understand the point of this second cache. From what you wrote in your last commit, it’s to save memory by only caching IDs, but… Sites are still stored in `SITE_CACHE`, so I don’t see how this saves memory.

Unless I totally missed your point, you can remove this cache, it should use more memory than earlier. Ditching this second cache also makes the code simpler, as we can directly fetch site objects from host+port strings.

Reply:
I was (perhaps naively) attempting to guard against the possibility of the original `SITE_CACHE` getting filled with duplicates in the event of some attack where ports could be spoofed for a matching domain, resulting in a Site object existing in memory for each port/hostname combination. In a genuine attack situation, I suppose it wouldn't make much difference in the grand scheme of things.

Reply:
The number of ports is finite, it’s 65535. So an hypothetical attack based on that would at max populate the dictionary with 65535 site objects × the number of hosts. This may take at max a dozen megabytes, so it would be a pretty lame attack.

That being said, this will never happen unless you did all these:
- configured your OS to accept connections on all ports
- configured your HTTP server to redirect by default all hosts and ports to a Wagtail instance
- configured Django with something like `ALLOWED_HOSTS = ['*']`

So this will not happen in any correctly configured server, and would actually need some extra work to allow any port through the HTTP server: by default they only listen to ports 80 & 443, as far as I know.

It’s possible though to have an infinite number of hosts if you setup wildcard subdomains and adjust `ALLOWED_HOSTS` accordingly. But in that event, you also need to have an huge number of `Site` instances in the database, so we can’t avoid caching each one of them. If a subdomain does not exist in the database, then we simply fetch the default one using the `'__default__'` key, so it doesn’t change memory usage.

Reply:
Yes, thanks for the insight. Like I said, it was a little naive. I'll re-simplify.